### ISSUE TITLE
Decouple voting functionality from MiniMe token

### ISSUE THREAD

**Issue by ptrwtts** [aragon/aragon-apps#388]
A lot of the Aragon functionality (Voting, Survey, etc) appears to rely on the `balanceOfAt` function found in the [MiniMe](https://github.com/Giveth/minime) token. This means that Aragon Apps only work with MiniMe tokens. It would be good if the checkpointing functionality was generalized, so that developers could easily add it to their tokens, and these tokens would be compatible with Aragon.

Suggested steps:

- [ ] Propose a token standard for "checkpointed" tokens
- [ ] Create a reference implementation that is compatible with libraries like Open Zeppelin
- [ ] Update Aragon Apps to work with any token that follows the standard (not just MiniMe)

We are open to helping with the standard / reference implementation, but wanted to first get feedback on whether this sounds like a good plan.

**Comment by sohkai** [comment]
This does sound like a good plan, it might be best to add an EIP for "snapshottable" tokens, based on the MiniMe design as a lot of projects have deployed MiniMe tokens.

Note that some apps, e.g. our Token Manager, requires more than just the snapshottable functionality. They also require the controller callbacks that's similar to the ERC-777 operators.

**Comment by ptrwtts** [comment]
Interesting. Which controller elements are used? The ability to generate/destroy tokens? These would also benefit from some sort of standard, although there's a lot more inconsistency in how deployed tokens have implemented it (MiniMe uses create/destroy, while Open Zeppelin uses mint/burn). 

Is it possible to design the apps so that they work with either generateTokens() or mint()?

As for "snapshottable" tokens, MiniMe is the only real example I've seen, so it makes sense to use it as the base for a standard, so that existing tokens are compatible.

**Comment by ptrwtts** [comment]
Thanks @sohkai, just noticed the link you added.

If we were to do our own implementation of balanceOfAt, how hard would it be to adapt Aragon apps to work with this? I'm basically trying to figure out if it's a battle worth fighting, or if we should just switch to MiniMe.

On a related note, I know the Aragon team have been vocal supporters of ERC-777. Is there any plans for a new token template that is both "snapshottable" and ERC-777 compatible?  

Basically I'm trying to get clarity on what the best path of action is for someone implementing a token over the next few months, and looking to be compatible with Aragon in the future. Cheers.

**Comment by izqui** [comment]
The giveth team was working on an implementation for an ERC777 'snapshottable' token here https://github.com/Giveth/yogatoken. I don't think it is updated to the latest ERC777 spec but I imagine they will work on it in the future. 

We are currently re-evalutating whether requiring all voting tokens to be MiniMe is a good idea going forward. A good alternative (that we are already using https://github.com/aragonlabs/plcr) is requiring users to stake a token (any ERC20) in order to gain voting rights. This would come at the expense of UX, but could allow for using any token as a governance token. The tokens could be unstaked at any point, unless it is decided otherwise when designing the governance mechanism (e.g. if voting in favor of a proposal, tokens become locked for x months). What are your thoughts on requiring users to stake before voting @ptrwtts?

**Comment by sohkai** [comment]
@ptrwtts 

> If we were to do our own implementation of balanceOfAt, how hard would it be to adapt Aragon apps to work with this?

This is hard to answer right now. Given that ANT is itself a MiniMe token, you can image we'll try hard to preserve compatibility with it, even if we decide to move to other strategies (e.g. staking).

> Is it possible to design the apps so that they work with either generateTokens() or mint()?

Potentially if the tokens either declare their interface or one of the interface EIPs becomes widely used... but for now neither of these are true.

Aside from minting / burning, we also use the controller functionality to control transfers sometimes.

**Comment by ptrwtts** [comment]
Staking is certainly a good way to achieve wide compatibility with tokens, although it's potentially a step backwards in UX from MiniMe. The main issues are that tokens a) are locked up and b) no longer appear in your wallet balance. But I think it could work if it used partial locks, so you can vote on multiple proposals in parallel, and followed a standard (maybe [ERC900](https://github.com/ethereum/EIPs/issues/900)?), so that wallets could reflect your staked balance.

**Comment by ptrwtts** [comment]
@sohkai / @izqui hi, just wondering if you've made any decisions around whether MiniMe will continue to be a requirement?

**Comment by luisivan** [comment]
From what I know, the current state of affairs is that we will implement a Staking app (contracts are done, we need a UI) so in order to vote, you need to stake the tokens first. That way, we would decouple the voting functionality from the token itself, and it will open the door for any kind of token to be used for voting

**Comment by bingen** [comment]
Hey @ptrwtts you can see some work on that checkpointing idea here: https://github.com/aragon/aragon-apps/pull/415
Plus some broader discussion about how to move forward with Staking that may be of your interest:
https://research.aragon.org/t/staking-locks-spec-v2/217

**Comment by sohkai** [comment]
@ptrwtts We've written up some of our research on this: http://blog.aragon.one/aragon-voting-part-1-ethereum-storage-time-machine/

**Comment by ptrwtts** [comment]
Bravo. 

The use of proofs is a great way to bring checkpointing to any token. And by the sounds of it, you're also planning to do aggregated submissions as a way to lower the cost per voter, which is also cool. These are the sorts of solutions we were thinking we would need to build _on top_ of Aragon, but baking them into the system and making it available to everyone makes a lot more sense.

Separately, EVM Storage Proofs sound really interesting. Are you planning to make your oracle system generic, so that the proofs can be used for other applications outside of Aragon / voting?

**Comment by sohkai** [comment]
@ptrwtts 

> Are you planning to make your oracle system generic, so that the proofs can be used for other applications outside of Aragon / voting?

We're still brainstorming use cases for the proofs, but we'd love to hear any particular areas you think this would apply to. The most immediate use case is checkpointing, of course :).

The gas cost requirements for sending the proof on-chain is somewhat prohibitive for cases where a simple SLOAD would work.

**Comment by sohkai** [comment]
Closing; there were efforts to alleviate this without forcing non-MiniMe tokens:

- [TokenWrapper](https://github.com/aragonone/voting-connectors/tree/master/apps/token-wrapper) (audited and usable on mainnet)
- [SVRP Voting](https://github.com/aragon/svrp) (research)
